### ISSUE TITLE
[Feature]: middleware before Remix processes request

### ISSUE THREAD

**Issue by a-type** [remix-run/remix#1587]
### What is the new or updated feature that you are suggesting?

Enjoying Remix so far, but there's one key feature I'd like to see for multi-tenant apps: the ability to intercept and modify an incoming request before handing it off to the core Remix runtime which is agnostic of the deployed platform.

The multi-tenant use case is to enable wildcard subdomains which route to a particular nested route structure within the app. For example, `foo.myapp.com/bar` could be rewritten and rendered as `myapp.com/_sites/foo/bar`, utilizing the routing structure within `/routes/_sites` to render the `foo` tenant's 'website' using the same codebase as the core app.

This behavior can't be achieved within `entry.server.ts`, since by the point that is called, Remix has already done up-front route matching on the original URL. Any userland code we could write to achieve this rewriting would have to be supplied to the server runtime directly.

Currently, the only working alternative which enables URL rewrite-based multi-tenancy is to modify the platform-specific deployment entrypoint (i.e. `/api/index.js` for Vercel, etc) to rewrite the URL before Remix gets it. But this only works in deployed environments (not locally), and would have to be rewritten for different deployment targets if you switch.

CC: https://github.com/remix-run/remix/discussions/1475

### Why should this feature be included?

- Multi-tenancy is very difficult to get right with Remix without it. You either have to include ballooning special-case logic within the routes themselves ("am I on a subdomain tenant site or the main site?" on every route) or customize your deployed platform code to modify the request beforehand and accept not being able to test subdomains on localhost.
- The ... _obvious competing framework_ has explicit support for [subdomain mult-tenancy](https://demo.vercel.pub/platforms-starter-kit) utilizing a special ['middleware' concept](https://vercel.com/docs/concepts/functions/edge-functions#middleware), roughly similar to what I've requested here.
- Subdomain multi-tenancy is increasingly common and a desirable feature for all sorts of sectors, whether it's a B2B dashboard, communication apps for teams, website builders, nocode tools, social network apps, or anything where a user would want to stake a claim on a section of your app.
- A request middleware could also prove more useful than just for multi-tenancy or URL rewriting at all. Perhaps you want to add or remove other parts of the request, like headers or cookies.

**Comment by esamattis** [comment]
Another important use for middleware is database connection pooling. We need to be able to ensure the database connections are released back to the pool at end end of the request even if an error is thrown in a loader/action.

So for now we need to use this `useDB(async db => {})` wrapper to ensure we don't accidentally run out of PostgreSQL connections. Where the `useDB()` is a try-finally wrapper like this:

```ts
function useDB(fn) {
    const db = POOL.get(); 
    try {
        await fn(db);
    } finally {
        db.release();
    }
}
```


This is OKish but it makes loaders and actions needlessly nested:

```ts
export const loader: LoaderFunction = async () => {
    await useDB(async (db) => {
        await db.query(...)
    });
};
``` 

It would be awesome to have a central place to extend the loader/action context so we could implement something like this:

```ts
export const loader: LoaderFunction = async ({ db }) => {
    await db.query(...)
};
```

**Comment by a-type** [comment]
@esamattis good use case. It seems like `entry.server.ts` `handleDataRoute` could _almost_ do what you need, but it only exposes modifying a response, not doing pre-request setup. There's still the problem of instantiating the pool and putting it into context.

This would be at least feasible to improvise in userland in the meantime, with a custom loader/action higher-order wrapper -

```ts
function dbLoader(loaderImpl: (event: Parameters<LoaderFunction>[0], db: DbPool) => ReturnType<LoaderFunction>): LoaderFunction {
  const loader: LoaderFunction = (event) => {
    return useDB((db) => {
      return loaderImpl(event, db);
    });
  }
  return loader;
}

// ...

export const loader = dbLoader(async ({ request }, db) => {
  await db.query(...);
});
```

**Comment by esamattis** [comment]
I think user land middleware should be now possible with `getLoadContext()` in the Express adapter which [gets access to the Node.js request and response](https://github.com/remix-run/remix/blob/db2c31f64affb2095e4286b91306b96435967969/packages/remix-express/server.ts#L60) just by listening on the `end` event of the response object. But that's going to be Express/node.js specific. 

> It seems like `entry.server.ts` `handleDataRoute` could _almost_ do what you need, but it only exposes modifying a response, not doing pre-request setup

Is it executed on loaders and actions too? Seems like it is always doing react rending 🤔 

> a custom loader/action higher-order wrapper

This is actually pretty good solution but I wonder whether does it get properly removed from the browser builds. 

Thanks for the ideas! I'll be investigating/testing these more on next week!

**Comment by esamattis** [comment]
> I wonder whether does it get properly removed from the browser builds.

Did some testing and **it does not get removed from the browser bundles under `public/build`** but it seems Remix does not remove remove even the plain `loader` and `action` code [like Next.js does for `getServerSideProps`.](https://next-code-elimination.vercel.app/) So I guess this is a separate issue (or feature?).

**Comment by a-type** [comment]
Hm, that's troubling (probably with a second issue). I'd hope that we could introduce composition functions over loaders and actions in userland without any problems.

**Comment by miraage** [comment]
I was looking for a graceful way to handle subdomain multi-tenant logic. Out of the box support would be much appreciated. 

Same goes for language/region/city detection. I can technically use a loader in root.tsx, but I think this logic is application-wide and doesn't necessarily belong to anything related to react.

**Comment by styxlab** [comment]
There is no need for a middleware concept for multi-tenancy apps (see my comment on #1337).

**Comment by a-type** [comment]
That will only work if route paths match exactly between subdomain situations and non-subdomain - i.e. `tenant.my.app/page` is the same page as `my.app/page`.

But if `tenant.my.app/page` is equivalent to `my.app/~tenants/[tenantName]/page` or some other structure, the solution is not as simple as just knowing the hostname - you also have to render an entirely different experience for route structures based on that host.

This ballooning complexity problem is what I was referring to in the original issue text. The only simple solution I can see is to rewrite `tenant.my.app` to `my.app/~tenants/[tenantName]` before Remix does anything at all. Then multi-tenancy is a 1-line feature tucked away into my middleware and a straightforward nested route quirk, not a sprawling concern spread across every route and loader in the codebase.

**Comment by styxlab** [comment]
@a-type Yes, I agree - in my use case, the routing structure does not change based on different tenants. The ability to perform rewrites in middleware would open more complex use cases and I would also like to see it. If you are on Cloudflare, you could dedicate a worker just for doing the rewrites. So there are workarounds, although I would prefer a native remix solution.

**Comment by kiliman** [comment]
One of the reasons why Remix doesn't support rewriting URLs is that the URL that it's processing must match the URL that is in the address bar. This is because React Router (which Remix is built on) does client-side routing so the server URL must match the client URL.

They would have to include URL rewriting in React Router itself.

**Comment by styxlab** [comment]
Maybe I don't understand the depth of the issue, but wouldn't it be possible to have a middleware/proxy before entering React router?
```
incoming request -> middlewar/proxy -> remix
```

**Comment by a-type** [comment]
The client-side routing definitely complicates things! Seems like the middleware may have to live in RR itself. However, that also means that even solutions like rewriting the incoming HTTP request in your lambda layer won't work, right? It would seem 

Remix doesn't have any solution for the kind of use case you see in NextJS' [Multi-Tenant demo](https://vercel.com/guides/nextjs-multi-tenant-application) using their middleware concepts. I'm not sure how NextJS middleware interacts with their client-side routing, but they've specifically showcased this kind of thing which at least validates that Vercel sees the need for this use case enough to commit effort to documenting it.

**Comment by dcramer** [comment]
Just to chime in here, this is also something that would be an appropriate entry point for functionality like Sentry to inject itself on the server. I think the pattern of having something like `middleware.server.tsx` and `middleware.client.tsx` could function here as its fairly easy know whats going on at that rate.

Since folks are contrasting to Next.js, I will say that their approach to middleware is good in some situations, but still limiting in others. For example, if you use Google's Identity-Aware Proxy it passes a signed JWT to every request which contains the authentication information.  In an ideal world an app (a Remix app) could parse this and bundle that information up in its hydration phase (middleware?). It could then retain that state in memory to understand if/who the authenticated session is. AFAIK in Next.js the only solution here is to utilize cookies to pass that information. There may be other approaches but it'd certainly be convulated.

Similarly in Sentry we'd want to inject something as a middleware that ideally could access things like routing information, HTTP headers, etc. allowing us to enrich data streams (both for error reporting and tracing). That enrichment behavior would be fairly distinct on the server, but would still be desirable on the client. So while it's certainly possible you could come up with an API that lets you write-once-run-everywhere middleware, I personally would be fine with having an implementation for each runtime.
